quicktill/card.py

```diff
@@ -24,53 +14,71 @@ class cardpopup(ui.dismisspopup):
     terminal) and whether there is any cashback on this transaction.
     """
     def __init__(self, reg, transid, amount, func, max_cashback, cashback_first,
-                 refund=False):
+                 ask_for_machine_id=False, refund=False):
         self.reg = reg
         self.transid = transid
         self.amount = amount
         self.func = func
         self.max_cashback = max_cashback
+        self.ask_for_machine_id = ask_for_machine_id
         cashback_in_use = max_cashback > zero
         h = 18 if cashback_in_use else 10
+        if cashback_in_use:
+            h = 18
+        else:
+            if ask_for_machine_id:
+                h = 12
+            else:
+                h = 10
         desc = "refund" if refund else "payment"
         ui.dismisspopup.__init__(
             self, h, 44, title="Card {} transaction {}".format(
                 desc, transid), colour=ui.colour_input)
-        self.addstr(2, 2, "Card {} of {}".format(desc, tillconfig.fc(amount)))
         if cashback_in_use:
             if cashback_first:
-                cbstart = 4
-                rnstart = 12
+                cbstart = 2
+                rnstart = 10
             else:
-                cbstart = 9
-                rnstart = 4
+                cbstart = 9 if ask_for_machine_id else 7
+                rnstart = 2
         else:
+            self.addstr(2, 2, "Card {} of {}".format(desc, tillconfig.fc(amount)))
```

Comment:
FWIW, I found this information to be redundant when cashback is enabled, so I only display it when cashback isn't displayed in favor of a less bloated dialog.

Reply:
It's there so that there's an obvious, consistent place for members of staff to look to see what figure they should be entering into the card terminal.  The UI flow on the card terminal is:

1. Staff member enters transaction amount
2. Staff member inserts customer's card
3. If card supports cashback, card terminal prompts for cashback amount
4. Staff member hands terminal to customer
5. Customer enters PIN and returns terminal to staff member
6. Card terminal prints merchant receipt (with receipt number at bottom, and cashback amount if required)
7. Staff member removes card from card terminal, enters cashback amount and receipt number into till, collects cashback from cash drawer if required.
7. Card terminal prints customer receipt; staff member returns card, receipt and cashback to customer

This is horrible because we don't have (semi-)integrated card terminals!  The flow with those would be:

1. Staff member hands terminal to customer
2. Customer inserts card, is prompted for cashback if supported
3. Customer enters PIN
4. Card terminal prints customer receipt, says "remove card and return terminal"
5. If cashback was selected, staff member collects cash and hands to customer

(With obvious shortcuts for contactless / no cashback)